src/main/java/com/google/sps/ChatWebSocket.java

```diff
@@ -59,9 +64,45 @@ public void onOpen(Session session) throws IOException {
     }
 
     @OnMessage
-    public void textMessage(Session session, String msg) {
-        System.out.println("Text message: " + msg);
+    public void textMessage(Session session, String msg) throws IOException {
+        JSONObject messageJson = new JSONObject(msg);
+
+        switch (messageJson.getString("type")) {
+            case "MSG_SEND":
+                chatMessage(session, messageJson);
+                break;
+            case "MAP_SEND":
+
+                break;
+            default:
+                System.out.println("Invalid message");
+        }
     }
+
+    private void chatMessage(Session session, JSONObject messageData)
+            throws IOException {
+        System.out.println("Text message: " + messageData.getString("message"));
+
+        List<Session> participants =
+            WebSocketHandler.getInstance().getRoomList(chatRoom);
+
+        JSONObject echoData = new JSONObject();
+        echoData.put("type", "MSG_RECV");
+        echoData.put("message", messageData.getString("message"));
+        echoData.put("uid", uid);
+        String echoJson = echoData.toString();
+
+        for (Session participant : participants) {
+            participant.getBasicRemote().sendText(echoJson);
+        }
+    }
+
+    private void mapMessage(Session session, JSONObject messageData) {
+        System.out.println("Map data: "
```

Comment:
Might think about removing if this isn't related to code functionality

Reply:
+1 For this case, it'd be appropriate to throw an `UnsupportedOperationException`